### PR TITLE
refactored ./src/types/path-set

### PR DESCRIPTION
**Pull request by natenichols**
## High Level Overview of Change
Refactored Hop, Path, and PathSet classes.

### Context of Change
Refactoring Hop, Path, and PathSet to extends serialized type and constructor parameters to be of type Buffer, to match how the base class is constructed. Moved from `makeClass()` -> `class`.

### Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Before / After
**Before:**
All classes were constructed from and stored `HopObject`, `Array<Hop>`, `Array<Path>` as members, and then used `toJSON` and `toBytesSink` members to serialize and deserialize. 

**After:**
`Hop`, `Path`, and `PathSet` each implement `from()`, `fromParser()`, and `toJSON()`. `Hop` is built `from`  a `HopObject`, `Path` is built `from` an `Array<HopObject>`, and `PathSet` can be constructed `from` an `Array<Array<HopObject>>`. Similarly, the `toJSON` method on `Hop` constructs a `HopObject`, and so forth. 

## Test Plan
All tests still pass, no tests added.

## Notes
Also, I ran `yarn lint`, and some changes to `UInt` and derived were not included in the last PR. No functionality was changed in any `UInt` or derived classes

### CODE REVIEW

File: src/types/path-set.ts [hunk 1]
```diff
@@ -1,120 +1,248 @@
...
+  issuer?: string;
+  account?: string;
+  currency?: string;
+}
+
+/**
+ * Serialize and Deserialize a Hop
+ */
+class Hop extends SerializedTypeClass {
+  constructor(bytes: Buffer) {
```
Comment (keefertaylor) [inline src/types/path-set.ts:32]: OOC: Can you just omit this and `new Hop(bytes)` will call super?

File: src/types/path-set.ts [hunk 2]
```diff
@@ -1,120 +1,248 @@
...
+  }
+
+  /**
+   * Create a Hop from a HopObject
+   * 
+   * @param value Either a hop or HopObject to create a hop with
+   * @returns a Hop
+   */
+  static from(value: Hop | HopObject): Hop {
+    if (value instanceof this) {
```
Comment (keefertaylor) [inline src/types/path-set.ts:43]: I'd prefer `Hop` to be explicit about the type you're checking for and I think it makes your code more readable too. 

The advantage to `this` (AFAICT) is that if you renamed the class you wouldn't need to change this line. I'm not sure that's an overwhelming reason to compromise readability. 

Let me know if you feel strongly.

File: src/types/path-set.ts [hunk 3]
```diff
@@ -1,120 +1,248 @@
...
+  /**
+   * Get the JSON interpretation of this hop
+   * 
+   * @returns a HopObject, an JS object with optional account, issuer, and currency
+   */
+  toJSON(): HopObject {
+    const hopParser = new BinaryParser(this.bytes.toString("hex"));
+    const type = hopParser.readUInt8();
+
+    const ret: HopObject = {};
```
Comment (keefertaylor) [inline src/types/path-set.ts:92]: nit: prefer `result` to `ret`

File: src/types/path-set.ts [hunk 4]
```diff
@@ -1,120 +1,248 @@
...
+   * Construct a Hop from a BinaryParser
+   * 
+   * @param parser BinaryParser to read the Hop from
+   * @returns a Hop
+   */
+  static fromParser(parser: BinaryParser): Hop {
+    const type = parser.readUInt8();
+    const bytes: Array<Buffer> = [Buffer.from([type])];
+
+    type & TYPE_ACCOUNT && bytes.push(parser.read(AccountID.width));
```
Comment (keefertaylor) [inline src/types/path-set.ts:77]: optional: 
I know you're just translating code, but I generally think it's more readable to write this as:
```js
if (type & TYPE_ACCOUNT) {
  bytes.push(parser.read(AccountID.width));
}
```

I prefer brackets but you could also do:
```js
if (type & TYPE_ACCOUNT) bytes.push(parser.read(AccountID.width));
```

if you wanted to be brief. 

Both of these strike me as more idiomatic and readable than the existing code, unless there's something magical going on that I don't understand.

Comment (keefertaylor) [reply src/types/path-set.ts:77]: This is also true on :92 so if you address this please be consistent.

File: src/types/uint-16.ts [hunk 5]
```diff
@@ -5,11 +5,13 @@ import { BinaryParser } from "../serdes/binary-parser";
  * Derived UInt class for serializing/deserializing 16 bit UInt
  */
 class UInt16 extends UInt {
-  protected static readonly width: number = 16 / 8 //2
-  static readonly defaultUInt16: UInt16 = new UInt16(Buffer.alloc(UInt16.width))
+  protected static readonly width: number = 16 / 8; //2
```
Comment (keefertaylor) [inline src/types/uint-16.ts:8]: Space after comment:

```suggestion
  protected static readonly width: number = 16 / 8; // 2
```

Comment (keefertaylor) [reply src/types/uint-16.ts:8]: Unless this is a linter reverting it?

File: src/types/path-set.ts (hunk 1 above)
Comment (natenichols) [reply src/types/path-set.ts:32]: Yup. Removing constructor from all three classes. `new` will just call the Base class

File: src/types/uint-16.ts (hunk 5 above)
Comment (natenichols) [reply src/types/uint-16.ts:8]: It was the linter that made the changes to `UInt` classes.

**Pull request merged**